keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/constraint/ImageExistsConstraintEvaluator.kt

```diff
@@ -50,17 +53,29 @@ class ImageExistsConstraintEvaluator(
     return image != null
   }
 
-  private fun findMatchingImage(version: String, vmOptions: VirtualMachineOptions): NamedImage? =
-    runBlocking {
-      log.debug("Searching for baked image for {} in {}", version, vmOptions.regions.joinToString())
+  private data class Key(
+    val account: String,
+    val version: String,
+    val regions: Set<String>
+  )
+
+  private val cache = cacheFactory
+    .asyncLoadingCache<Key, NamedImage>(cacheName = "namedImages") { key ->
+      log.debug("Searching for baked image for {} in {}", key.version, key.regions.joinToString())
       imageService.getLatestNamedImageWithAllRegionsForAppVersion(
         // TODO: Frigga and Rocket version parsing are not aligned. We should consolidate.
-        AppVersion.parseName(version) ?: throw SystemException("Invalid AMI app version: $version"),
-        defaultImageAccount,
-        vmOptions.regions
+        appVersion = AppVersion.parseName(key.version)
+          ?: throw SystemException("Invalid AMI app version: ${key.version}"),
+        account = key.account,
+        regions = key.regions
       )
     }
```

Comment:
Reintroducing a cache directly in the constraint evaluator because we want to cache the results that have been filtered by `getLatestNamedImageWithAllRegionsForAppVersion` to ensure all desired regions are present.